File: Decode_To_String/decode_to_string.py
class Solution:
    def decodeString(self, s: str) -> str:     
        if not s:
            return ""

        def decode(s: str, index: int) -> (str, int):
            num = 0
            localString = ""
            while index < len(s):
                char = s[index]

                if char.isdigit():
                    num = num * 10 + int(char)
                                        
                elif char.isalpha():
                    localString += char

                elif (char == '['):
                    inner, index = decode(s, index +1)
                    localString += inner * num
                    num = 0
                elif (char == ']'):
                    return localString, index


                index += 1

            return localString, index

        regex, _ = decode(s, 0)
        return regex

File: Decode_To_String/decode_to_string.cpp
#include <string>
#include <stack>
#include <cctype>


class Solution {
public:
    string decodeString(string s) {
        stack<string>  strStack;
        stack<int> numStack;

        string curr_str = "";
        int num = 0;

        for (int i = 0; i < s.length(); i++ ){
            if (isdigit(s[i])) {
                num = num * 10 + (int(s[i])-'0'); 
            }
            else if ((s[i]) == '[') {
                strStack.push(curr_str);
                numStack.push(num);
                curr_str = "";
                num = 0;
            }
            else if (s[i] == ']') {
                int repeat = numStack.top(); // return by reference
                numStack.pop();

                string prev_str = strStack.top();
                strStack.pop();

                string temp = "";
                for (int j = 0; j < repeat; j++) {
                    temp += curr_str;
                }
                curr_str = prev_str + temp;
            }
            else {
                curr_str += s[i];
            }
        }
        return curr_str;
    }
};
